### DISCUSSION TITLE
a common class, imageLoader

### DISCUSSION THREAD

**Discussion by rageworx** [fire-eggs/YAIV#44] [Ideas]
I'm thinking a new class  like imageLoader.
- This class just has one instance ( singletone ).
- Each image automatically sensed, and generated for Fl_RGB_Image and a flag for timered image for GIF, PNG.
- In case of timerd image returns with Fl_Timered_Image instead of animation images.
- Fl_Timered_Image has own timer and event handler for update its own image to exposed.
- Viewer will connected handling event with Fl_Timered_Image, and it will update current image with new image for animation event.
- options : meta informations ( like EXiF )

It will makes any viewer for more simple, and easy to use, easy to implement all features like most of image viewer.
I will start this idea as an new project for FLTK imaging, for Make Image O' HDR.
How about this ?

**Comment by fire-eggs** [comment]
Very nice! Its been on my to-do list for a while but keeps getting pushed back.

**Comment by rageworx** [comment]
Now on progress in private repo.

![image](https://user-images.githubusercontent.com/6948225/121334601-c2594c00-c954-11eb-91e4-928c5871538e.png)

I will be share this repo within a month or weeks.
And .. there's some accident to me about a few days ago.
My right knee has some problem, so I can't sit or lay down anywhere for hours - even office too...
( so it will be took more days longer for doing something whatever it is .. )

some info, libfl_imgldr requires libexif, libturbojpeg for getting EXiF meta data.
I will be back here soon !

**Reply by fire-eggs** [reply]
Thanks for the update! You're tackling two things on my list: loader and metadata. Excellent!

Sorry to hear about your injury! You take care and don't feel you need to rush anything for my sake - I've got plenty on my to-do list !

**Comment by rageworx** [comment]
Currently libfl_imgldr reads exif as well.
![image](https://user-images.githubusercontent.com/6948225/121859958-52233f80-cd33-11eb-93b9-25feebd64473.png)

**Comment by rageworx** [comment]
now meta information recognizes orientation and automatically apply flip and rotation if 'applymetainfo' is true.
```
Fl_RGB_Image* fl_imgldr::loadimage( const uint8_t* buff, size_t bufflen, bool applymetainfo )
```

**Reply by rageworx** [reply]
Orientation from libexif, to flip and rotate degree ...
```
static void fl_imgldr_orientation_str_to_degree( const char* ostr, unsigned& deg, unsigned& flip )
{
    // following libexif.
    // ref. https://www.impulseadventure.com/photo/exif-orientation.html
    // and, https://www.impulseadventure.com/photo/images/orient_flag.gif
    const char* otables[9] =
    { "",               /// not used.
      "Top-left",       /// 1 == 0'
      "Top-right",      /// 2 == Flip vertical + 0'
      "Bottom-right",   /// 3 == 180'
      "Bottom-left",    /// 4 == Flip vertical + CW 180'
      "Left-top",       /// 5 == Flip vertical + 270'
      "Right-top",      /// 6 == 90'
      "Right-bottom",   /// 7 == Flip vertical + 90'
      "Left-bottom",    /// 8 == 270'
    };
    
    int fpos = -1;
    for ( size_t cnt=1; cnt<9; cnt++ )
    {
        if ( strcmp( ostr, otables[cnt] ) == 0 )
        {
            fpos = (int)cnt;
            break;
        }
    }
    
    switch( fpos )
    {
        default:
        case 1:
            deg = 0;
            flip = 0;
            break;

        case 2:
            deg = 0;
            flip = 1;
            break;
            
        case 3:
            deg = 180;
            flip = 0;
            break;
        
        case 4:
            deg = 180;
            flip = 1;
            break;
            
        case 5:
            deg = 270;
            flip = 1;
            break;
            
        case 6:
            deg = 90;
            flip = 0;
            break;
            
        case 7:
            deg = 90;
            flip = 1;
            break;
            
        case 8:
            deg = 270;
            flip = 0;
            break;
    }
}
```
